cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceService.java

```diff
@@ -0,0 +1,215 @@
+  private static final String NAMESPACE_ELEMENT_TYPE = "Namespace";
+
+  private final Store store;
+  private final PreferencesStore preferencesStore;
+  private final DashboardStore dashboardStore;
+
+  @Inject
+  public NamespaceService(StoreFactory storeFactory, PreferencesStore preferencesStore, DashboardStore dashboardStore) {
+    this.store = storeFactory.create();
+    this.preferencesStore = preferencesStore;
+    this.dashboardStore = dashboardStore;
+  }
+
+  /**
+   * Initializes and starts namespace service.
+   * Currently only creates the default namespace if it does not exist.
+   * In future, it could potentially do other tasks as well.
+   */
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting namespace service");
+    ensureDefaultNamespaceExists();
+  }
+
+  /**
+   * Asynchronously ensures that the default namespace exists, with retries.
+   * Need a retry mechanism since the ensure operation requires the dataset service to be running.
+   * Hence this method ensures that the default namespace exists in a separate thread, with multiple retries
+   */
+  private void ensureDefaultNamespaceExists() {
+    ListenableFuture<Void> listenableFuture = asyncCreateDefaultNamespace();
+    Futures.addCallback(listenableFuture, new FutureCallback<Void>() {
+      @Override
+      public void onSuccess(Void aVoid) {
+        // No-op, since either default namespace already existed, or its creation was successful
+        LOG.info("Successfully ensured existence of default namespace.");
+      }
+
+      @Override
+      public void onFailure(Throwable throwable) {
+        LOG.warn("Failed to ensure existence of default namespace.");
+        // now keep retrying
+        if (isRunning()) {
+          final FutureCallback<Void> callback = this;
+          // Retry in 2 seconds. Shouldn't sleep in this callback thread. Should start a new thread for the retry.
+          Thread retryThread = new Thread("retrying-default-namespace-ensurer") {
+            @Override
+            public void run() {
+              try {
+                TimeUnit.SECONDS.sleep(2);
+                LOG.info("Retrying to ensure that default namespace exists.");
+                Futures.addCallback(asyncCreateDefaultNamespace(), callback);
+              } catch (InterruptedException e) {
+                LOG.warn("Default namespace ensurer retry thread interrupted.");
+              }
+            }
+          };
+          retryThread.setDaemon(true);
+          retryThread.start();
+        }
+      }
+    });
+  }
+
+  /**
+   * Asynchronously calls #createDefaultNamespace to create the default namespace
+   */
+  private ListenableFuture<Void> asyncCreateDefaultNamespace() {
+    ListeningExecutorService executorService = MoreExecutors.listeningDecorator(
+      Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("default-namespace-ensurer")));
```

Comment:
Also for case as simple as this, you can just have a new Thread and do the retry in that thread if it fails. No need to use Future and future callback.